What's happening in Chrome Extensions - July 2023 edition (#6787)
GoogleChrome/developer.chrome.com: site/en/blog/extension-news-july-2023/index.md

---
layout: 'layouts/blog-post.njk'
title: What's happening in Chrome Extensions?
description: >
 An overview of the changes that have happened so far this year in the Chrome Extension Platform. Plus exciting upcoming new extension features developers can look forward to.
date: 2023-07-13
authors:
 - amysteam
tags:
 - extensions-news
hero: 'image/BhuKGJaIeLNPW9ehns59NfwqKxF2/jzaO8VCyWJRXKgeIe68T.png'
alt: >
 What's happening in Chrome extensions?
---

So far, 2023 has been a busy year in the world of Chrome extensions. Your valuable feedback has allowed us to improve the extension platform and our documentation. We also continue collaborating with other browser vendors in the [WebExtensions Community Group][wecg] so that extension APIs work more consistently across browsers. 

Hi, I'm Amy, a technical writer for the Chrome extension team. In this post, I'm excited to share with you a few changes that the Chrome extension team has worked on during the first half of this year and what upcoming features will be released later this quarter. Let's get started!

{% Aside %}

> But if you're like me (TL;DR), feel free to jump to the [Coming soon...][sec-coming-apis] section for new extension features you can start looking forward to 🙂.

Author: The original text was changed.

All extension events now restart the extension service worker's idle timer. In Chrome 110, the hard five-minute maximum lifetime was removed for extension service workers. Also, messages to [native applications][doc-native-msg] and [messages within the extension][doc-messages] restart the idle timer. Read more about it in [The extension service worker lifecycle][sw-lifecycle] article.

### More API launches {: #apis-others }

- **Action API**: Starting Chrome 110, you can customize the badge text with [`setBadgeTextColor`()][action-set-color] and [`getBadgeTextColor()`][action-get-color]. Also, [`isEnabled()`][action-isenabled] allows you to check if the action is enabled for the current tab. 
- **Commands API**: The bug where extension shortcuts, declared in the manifest under [`"commands._execute_action"`][commands-execute-action], would not persist during conversion to MV3, was fixed in [Chrome 111][bug-fix-commands].
- **Downloads API**: The default downloads UI in Chrome has moved from a shelf at the bottom to the right side of the omnibox. To disable this behavior, you can use [`downloads.setUiOptions()`][downloads-setui] which replaces `setShelfEnabled()`.

> - **History API**: [`chrome.history.getVisits()`][history-get-visits] and [`chrome.history.search()`][history-search] also return data from other devices that have been synced to the local history database. This may result in more history entries and higher visit counts. `isLocal` was added to [`VisitItem`][history-visititem] in Chrome 115 to be able to filter by local visits only.

Author: Did you mean Stable? 115 Beta hit on [May 31st](https://chromiumdash.appspot.com/schedule)

> - **History API**: [`chrome.history.getVisits()`][history-get-visits] and [`chrome.history.search()`][history-search] also return data from other devices that have been synced to the local history database. This may result in more history entries and higher visit counts. `isLocal` was added to [`VisitItem`][history-visititem] in Chrome 115 to be able to filter by local visits only.

Reviewer: Yes.

- **Storage API**: on Chrome 112 the [`chrome.session`][storage-session] storage size was increased to 10MB. Then [`chrome.local`][storage-local] storage size was matched in Chrome 114.

### Coming soon... {: #coming-soon-apis }

Upcoming Chrome versions will introduce many features to make it easier for extensions to migrate to Manifest V3. For a list of MV3 migration related upcoming changes, check out our [known issues page][mv3-known-issues]. Additionally, we plan to add the following features:

- **DeclarativeNetRequest API**: The default value for the [isUrlFilterCaseSensitive][dnr-url-case] property will change to `false`. See the [WECG][wecg-dnr-url-case] thread.
- The **File Handling API** will allow ChromeOS extensions to open files with specified MIME types and file extensions. This feature is currently [behind a flag][wn-file-handling].
- **Runtime API**: We are releasing [`runtime.getContexts()`][runtime-getcontexts] to replace `extension.getViews()`, which is deprecated. This will allow extensions to determine if an extension page, like the side panel or offscreen document is open. See the [WECG][wecg-get-contexts] proposal.

> - **Service workers**: We're adding strong keep-alives to Chrome APIs that display a user prompt: [`permissions.request`][permission-request], [`desktopCapture.chooseDesktopMedia()`][desktop-capture], [`identity.launchWebAuthFlow()`][identity-launchwebflow], and [`management.uninstall()`][management-uninstall].

Reviewer: ```suggestion
- **Service workers**: We're adding strong keep-alives to Chrome APIs that display a user prompt: [`permissions.request()`][permission-request], [`desktopCapture.chooseDesktopMedia()`][desktop-capture], [`identity.launchWebAuthFlow()`][identity-launchwebflow], and [`management.uninstall()`][management-uninstall].
```

documentation issues on [developer.chrome.com][github-dcc].

### Highlights {: #doc-highlights }

- The new [MV3 Migration][mv3-migration] section provides practical ways to convert Manifest V2 extensions to Manifest V3.
- The [Extension service workers][sw-explainer] guide provides detailed information on extension service workers topics. These include how they are registered and updated, what the lifecycle looks like, how imports work, and more.
- The [Handle events with service workers][sw-tut] tutorial teaches the basics of extension service workers. It builds an omnibox extension that gives you quick access to extension API reference pages.

### More updates {: #doc-others }

- [Using Google Analytics 4][tut-ga4] demonstrates how to track the usage of your extension popup and service worker events.
- [Using geolocation][tut-geo] shows how to obtain the geographical location of the extension using the Offscreen API.
- [Audio recording and screen capture][tut-capture] teaches how to capture audio and video from tabs, windows, or screens using `chrome.tabCapture` and `navigator.mediaDevices.getDisplayMedia()` APIs.

> - We added new debugging tips to the [Debugging extensions][tut-debug] guide.

Reviewer: ```suggestion
- We've added new debugging tips to the [Debugging extensions][tut-debug] guide.
```

> - We've updated the [Permission Warnings Guidelines][guide-perm-warn] to make it easier to understand how permission warnings work and how you can provide a better user experience. Also, there are practical ways to check what warnings the user will see.

Reviewer: ```suggestion
- We've updated the [Permission warnings guidelines][guide-perm-warn] to make it easier to understand how permission warnings work and how you can provide a better user experience. Also, there are practical ways to check what warnings the user will see.
```

- Our team and contributors have also added new Manifest V3 extension samples: [WASM in extensions][gh-wasm], [Scripting API demo][gh-scripting], [Side Panel API cookbooks][gh-sidepanel], and [DeclarativeNetRequest API samples][gh-dnr]. You can explore other extension samples in our [GitHub samples repo][gh-ext-samples].

### Coming soon... {: #coming-soon-docs }

- How to migrate remote hosted code to Manifest V3.
- How to run automated tests for Chrome extensions.
- Improved Declarative Net Request guidance.
- Improvements to the content script explainer.

## 💡 Did you know? {: #tips }

Before we wrap up, I wanted to share a couple of useful tools and insights:

> - In Chrome 115, you can access devices via [WebHID from your Chrome extension][gh-webhid].

Reviewer: This only works because of a hack that keeps service workers alive. That's why Devlin had me postpone documenting it. I wouldn't mention this until it's done for real.

> - [Puppeteer][puppeteer] now supports testing in headless mode with [Chrome's headless "new"][chrome-headless].

Reviewer: headless "new"?

> - With the [Extension Update Testing Tool][gh-update-tool], you can check what permission warnings are triggered when permissions change in the manifest. This way, you can experience the update process as a user would. This is important because some permissions might disable the extension until the user grants access again.

Reviewer: ```suggestion
- With the [Extension Update Testing Tool][gh-update-tool], you can check what warnings are triggered when permissions change in the manifest. This way, you can experience the update process as a user would. This is important because some permissions might disable the extension until the user grants access again.
```

- We've added new debugging tips to the [Debugging extensions][tut-debug] guide.
- We've updated the [Permission warnings guidelines][guide-perm-warn] to make it easier to understand how permission warnings work and how you can provide a better user experience. Also, there are practical ways to check what warnings the user will see.
- Our team and contributors have also added new Manifest V3 extension samples: [WASM in extensions][gh-wasm], [Scripting API demo][gh-scripting], [Side Panel API cookbooks][gh-sidepanel], and [DeclarativeNetRequest API samples][gh-dnr]. You can explore other extension samples in our [GitHub samples repo][gh-ext-samples].

### Coming soon... {: #coming-soon-docs }

- How to migrate remote hosted code to Manifest V3.
- How to run automated tests for Chrome extensions.
- Improved Declarative Net Request guidance.
- Improvements to the content script explainer.

## 💡 Did you know? {: #tips }

Before we wrap up, we wanted to share a couple of useful tools and insights:

> - In Chrome 115, you can access devices via [WebHID from your Chrome extension][gh-webhid].

Author: ```suggestion
- Chrome started work on supporting [WebHID][gh-webhid]; you can start playing around with the API starting Chrome 15 (but be aware that it's still WIP).
```
@jpmedley WDTY?

> - [Puppeteer][puppeteer] now supports testing in headless mode with [Chrome's headless "new"][chrome-headless].

Author: ```suggestion
- [Puppeteer][puppeteer] now supports testing in headless mode using `--headless=new` . Read more about it on [Chrome's headless mode upgrade][chrome-headless] blog post.
```
